Fix "dyld: Library not loaded" problems on macosx.

vinklein opened the pull request:
set a "--prefix" when calling configure for libflint mpfr and mpir
add an "-install_name" flag for libflint
define an install path for mpfr and mpir in .build_dependencies file

Commit:
Fix "dyld: Library not loaded" problems on macosx.

set a "--prefix" when calling configure for libflint mpfr and mpir
add an "-install_name" flag for libflint
define an install path for mpfr and mpir in .build_dependencies file

vinklein commented:
https://travis-ci.org/vinklein/flint2/builds/368176636

videlec commented:
Pour le `absolute_path` est-ce que tu peux vérifier que la commande suivante fait le boulot sous Max OSX?
```
$ readlink -f REPO
```
Si c'est le cas, change toutes les occurrences de `absolute_path` en `readlink -f`.

videlec commented:
> https://travis-ci.org/vinklein/flint2/builds/368176636

cool :-)

videlec commented on .travis.yml:
```diff
@@ -20,7 +20,7 @@ compiler:
 install:
   - ./.build_dependencies
-  - ./configure --with-mpir=mpir-2.7.2 --with-mpfr=mpfr-4.0.0
+  - ./configure --with-mpir=local --with-mpfr=local --prefix=`pwd`
```
Please, don't use a relative paths here. Why not installing in `local` as the others library are?

videlec commented:
Ca ne règle pas complètement le problème. On veut pouvoir faire un `make check` sans faire un `make install` (typiquement on veut vérifier que ça marche avec les droits d'un utilisateur avant d'installer avec les droits du super utilisateur). C'est-à-dire: il est *interdit* d'utiliser `--prefix=pwd`. Une possibilité est de faire ce que suggérait Bill Hart: avoir une procédure de "post install" pour redéfinir le install_name proprement.

Par ailleurs, dans sa version précédente le Makefile supportait des versions compilées de gmp/mpfr mais non installées... je ne sais pas trop quoi en penser.

vinklein commented:
`
$ readlink -f REPO
`
Ca ne marche pas sous osx.

Commit:
Change libflint.dylib install_name after `make install`

The install name used for the build is consistent with
the build directory. That way the tests (launched by `make check`) don't
need a `make install`.

After a `make install` the install name of the freshly installed
libflint.dylib is modified with `install_name_tool -id`

Commit:
MacOS / travis.yml : display install name of ...

...libflint.dylib after `make install`.

Makefile.in: Fix missing tab char

vinklein commented:
En voilà une qui marche.
Sous travis j'ai rajouté un make install et un oltool pour mac, pour pouvoir checker visuellement que le `install_name_tool -id` fait le boulot attendu. 

Une autre solution au cas ou le if `[[ "$(OS)" = "Darwin" ]]` ne plait pas, ca serai de définir une variable `POST_INSTALL_CMD` dans le `configure` et dans laquelle on passe la commande install_name_tool. On laisserai cette commande à vide pour les cas non Darwin.
C'est pas forcément plus propre dans le sens ou il faudrai en plus définir `LIBDIR` qui n'est actuellement pas connue de `configure` (Elle n'apparait que dans `Makefile.in` actuellement).

Commit:
.build_dependencies : Remove whitespace

videlec commented on .travis.yml:
```diff
@@ -20,10 +20,12 @@ compiler:
   - ${MAKE} QUIET_CC=
   - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then otool -L libflint.dylib; fi
   - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then ldd libflint.so; fi
+  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then make install; fi
```
Why do you need to run `make install` on osx? This should not be needed for `make check`.

vinklein replied:
to run `install_name_tool -id` it's not required for the `make check` indeed.

videlec reviewed (commented):
I am ok with the if in the Makefile (that is what I initially suggested upstream).

videlec commented:
En effet, ça marche sans https://travis-ci.org/videlec/flint2/builds/369255659

videlec commented:
Est-ce que tu as une idée pour tester le install_name hook (post `make install`) dans travis?

videlec commented:
J'ai fait un script de check post install (qui essaye de compiler un programme minimal).

vinklein commented:
Le `make install` n'était pas pour le `make check` mais juste pour lancer le `install_name_tool`. 
Pour que ce soit plus clair j'aurai sans doute du placer les deux dernière lignes de `-install` sous `-script` après le `make check`.
Plutot que compiler un programme de test tu peu aussi linker un des programmes de test existant (l'un de ceux créés et lancé par `make check` sur la librairie installé avec un `install_name_tool -change "ancien_chemin_libflint.dylib" "nouveau_chemin_libflint.dylib" testprog`

videlec commented:
Tu as essayé la commande `make install`? Sur travis ça donne
```
if [[ "Linux" = "Darwin" ]]; then \
		install_name_tool -id "/home/travis/build/videlec/flint2/local/lib/libflint.so.13.5.2" "/home/travis/build/videlec/flint2/local/lib/libflint.so"; \
	fi
/bin/sh: 1: [[: not found
```

vinklein commented:
Dit moi si tu pense que vérifier que le `install_name_tool -id` ne plante pas est inutile. Dans ce cas je vire les modifications du travis.

vinklein commented:
Bizarre ... Oui bien sur qu'elle a été testé .... mais sur mac

videlec commented:
Dans tes rapports de log à https://travis-ci.org/vinklein/flint2, la commande n'est jamais lancée (ou je me trompe).

videlec commented:
ha non, je dis des bêtises

videlec commented:
mais pas que... à https://api.travis-ci.org/v3/job/369274625/log.txt le make install plante.

vinklein commented:
sisi elle passe 
```
0.00s$ if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then otool -L local/lib/libflint.dylib; fi
local/lib/libflint.dylib:
	/Users/travis/build/vinklein/flint2/local/lib/libflint-13.dylib (compatibility version 13.5.0, current version 13.5.2)
	/Users/travis/build/vinklein/flint2/mpfr-4.0.0/../local/lib/libmpfr.6.dylib (compatibility version 7.0.0, current version 7.0.0)
	/Users/travis/build/vinklein/flint2/mpir-2.7.2/../local/lib/libgmp.16.dylib (compatibility version 23.0.0, current version 23.2.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
```

vinklein commented:
Bon je vais resortir le portable pour touver une syntaxe qui ne casse pas sous linux

videlec commented:
pourquoi tu n'as pas utilisé `if [ "$(OS)" -eq "Darwin" ]`?

vinklein commented:
J'ai fait plein d'allés retours la dessus. Tant qu'on y est tu veux qu'on déplace le `make install` après le `make check` ou pas ?

Commit:
fix if [ "$(OS)" = "Darwin" ]; for linux

move make install after make check

videlec commented:
Oui. C'est indispensable (je l'ai fait dans ma branche).

videlec commented:
> J'ai fait plein d'allés retours la dessus.

Et pourquoi tu as choisi l'autre syntaxe ?

vinklein commented:
> Et pourquoi tu as choisi l'autre syntaxe ?

Mais parceque celle là marchai ( Sous mac ). En gros j'ai eu quelques bugs de syntax à cet endroit là donc j'ai essayé plusieurs façons de faire et évidement le problème de venais pas de `[` vs`[[`

videlec commented:
> Et pourquoi tu as choisi l'autre syntaxe ? 
>> Mais parceque celle là marchai ( Sous mac )

Ca marche sous bash mais pas sous sh (je pense pas que ce soit un problème de plateforme). Mais c'est curieux, il devrait y avoir bash sur travis. Si tu regardes dans le Makefile, tous les `if` utilisent la syntaxe `[ condition ]`.

videlec commented:
Mon script post-install fonctionne :-) https://travis-ci.org/videlec/flint2/builds/369290266

videlec commented:
(et du coup ton hook avec le install_name aussi :-)

vinklein commented:
Classe ! Mes modifs sont a environ 15 min de test donc je met à jour la PR dans 10-20 min si tout va bien.

videlec commented:
t'embête pas. j'ai fait les modifs dans ma branche (j'ai récupéré tes commits).

vinklein commented:
e8a5a389c605389ca88f212fae26b51987c27f20 ? Bon en même temps y'as rien de sorcier.

vinklein commented:
Ben voilà ils sont fini, et çà passe https://travis-ci.org/vinklein/flint2/jobs/369297169.

vinklein commented:
A ok c'est la branche test_travis. Je te laisse clore la PR.

videlec commented:
ben tu m'as pas laissé clore la PR :-)

vinklein commented:
Et oui je sais pas ... un réflexe peut être : )
